<!--- Mohammad Idrees Bhat | Tech Skills Trainer | AI/ML Consultant --->

<div style="background-color: #add8e6; padding: 10px; height: 70px; border-radius: 15px;">
    <div style="font-family: 'Georgia', serif; font-size: 20px; padding: 10px; text-align: center; position: absolute; right: 20px;color: #000000ff;">
        Mohammad Idrees Bhat<br>
        <span style="font-family: 'Arial', sans-serif;font-size: 12px; color: #000000ff;">Tech Skills Trainer | AI/ML Consultant</span>
    </div>
</div>

<h1 style=" background-color: #002147; color: White; padding: 30px; text-align:center"> Python Programming Language (Part 4) </h1>

<div style="background-color: lightgreen; color: black; padding: 10px;">
    <h1> Functions
</h1> </div>

Programming paradigms refer to the fundamental styles or approaches to programming that dictate how code is structured and organized. Python supports multiple programming paradigms, making it a versatile language. 

<div style="background-color: grey; color: black; padding: 10px;">
    <h4><b>AGENDA</b> <p><p>
Functions in Python <p><p>

Types of programming
1. Imperative Programming <p><p> 
2. Procedural Programming <p>
3. Object-Oriented Programming (OOP) <p>
4. Functional Programming <p>
5. Declarative Programming   
</h4> </div>

<div style="background-color: lightgreen; color: black; padding: 4px;">
    <h4> Functions
</h4> </div>


#### What is a Function?
A function is a reusable block of code that performs a specific task. Functions help organize code, make it easier to read, and allow you to reuse code without rewriting it.

A function is like a small machine or a recipe.

- You:

- Give inputs (called parameters or arguments),

- The function does some work,




### Defining a Function
You define a function using the `def` keyword, followed by the function name and parentheses. Inside the parentheses, you can specify **parameters** (inputs) for the function.

### Basic Structure of a Function
```python
def function_name(parameters):
    # Code block to execute
    return result


#### Types of functions

Two types:

- Perform a task
- Calculate and Return a value 

In [3]:
# does a task

def greet(name):
    print(f"Hi {name}")


# returns
def get_greeting(name):
    return f"Hi {name}" # returns a value

message = get_greeting("Idrees") # message object stores a value this time
print(message)

Hi Idrees


In [ ]:
# Simple Function

def add_numbers(a, b): # parameters # This function takes two numbers and returns their sum
    result = a + b  # Add the two parameters
    return result    # Return the result of the addition

# Call the function with arguments 5 and 3
sum_result = add_numbers(5, 3) # arguments
print(sum_result)  # Output: 8

8


Keyword Arguments and Default parameters

- So using kwargs, order of passing arguments does not matter
- code more readable
- you can set up default values

In [2]:
# Function with Default Parameters
# default parameters are values set for parameters in function defintion

def greet(name, greeting="Hello"): # defines a function greet with a default parameter greeting.
    # whatever you write here belongs to this function
    print("heyo!!!")
    message = f"{greeting}, {name}!"  # Create a greeting message
    return message                    # Return the greeting message

# add two line breaks for pep8 standard for cleaner code
greet() # throws an error - missing 1 required positional argument: 'name'

# save what the function returned ina variable
greeting_message1 = greet("Faizan", "Heya There!!!")
print(greeting_message1)              

# Call the function with the default greeting
greeting_message2 = greet("Samuel")      # note that we are not passing in the greeing here
print(greeting_message2)              

# Call function with the keyword arguments
greeting_message3 = greet(greeting="Ahoy mate!!!",name = "Idrees")
print(greeting_message3)

heyo!!!
Heya There!!!, Alice!
heyo!!!
Hello, Bob!


#### *args - Variable number of arguments

Sometimes we wish to have functions that can take as many inputs as needed for a situation
- eg. 

In [4]:
def multiply(x,y):
    return x*y

multiply(3,2) # 6 - works

# but what about multiple parameters
#multiply(2,3,4,5) # error - won't work because more arguments that required

# let's try this:

def multiply(*numbers): # use a plural name to indicate many parameters
    print(numbers)

multiply(3,4,2,3,5)
# so multiply(numbers) becomes a tuple
# we get a tuple, which is an iterable!!
# we cannot modify this tuple tho

(3, 4, 2, 3, 5)


In [5]:
# we can then modify the function to return an actual value too
# which we can store in variables

def multiply(*numbers):
    total = 1
    for number in numbers:
        #number = number * total
        total *= number

    return total

print(multiply(3,4,2))

24


In [8]:
# ANother function with Variable Arguments

def sum_all(*numbers): # allows the function to accept any number of arguments as a tuple.
    
    total = 0  # Initialize a variable to hold the total sum
    for num in numbers:  # Iterate over each number in the arguments
        total += num  # Add each number to the total
    return total  # Return the total sum

# Call the function with multiple arguments
total_sum = sum_all(1, 2, 3, 45, 4, 5)
print(total_sum) 

60


#### Exercise
- Short examples of how functions are used

In [ ]:
# A function to square a number

def square(n):
    return n * n

In [ ]:
# A function to check even number

def is_even(n):
    return n % 2 == 0

In [ ]:
# A function to compute area of a circle
# (Use π ≈ 3.14)

def area_of_circle(r):
    return 3.14 * r * r

In [ ]:
###

### Lambda Functions – Tiny One-Line Functions
- What is a lambda?

- A lambda function is a small, anonymous (no name) function.

- Used when you need a quick one-line function.

```python
lambda parameters: expression


In [ ]:
# Square two numbers
square = lambda x: x * x
print(square(5))  # 25

# add two numbers 
add = lambda a, b: a + b
print(add(2, 3))  # 5

Exercise

- Create a function to double a number using lambda:
- Create max_of_two using lambda and a conditional expression:


In [9]:
double = lambda x: x * 2
print(double(10))  # 20

max_of_two = lambda a, b: a if a > b else b
print(max_of_two(5, 9))  # 9

20
9


In [ ]:
c = lambda a,b : a + b

c(2,3)


SyntaxError: invalid syntax (3935171945.py, line 3)

### Recusrion - function calling itself 
- Recursion is when a function calls itself.

- You break a big problem into smaller versions of the same problem.

- But one must do it in a controlled way

- Two key parts:

    - Base case → when to stop. 
        - When the problem is so small that you can answer it directly.
        - This stops the recursion.

    - Recursive case → do a small part, then call the function again.

```python
def function():
    if (a very small case):
        return answer
    else: 
        return (function with a smaller input)

In [40]:
# create a countdown function

def countdown(n):
    # base condition
    if n == 0:
        print("Done!!")
        return n
    # problem to solve, and recursion
    print(n)
    countdown(n-1)

countdown(6)

    

6
5
4
3
2
1
Done!!


Example:

- To sum numbers from 1 to 5:

    - Take 5, then add the sum of 1 to 4.

    - To get sum of 1 to 4, take 4, then add sum of 1 to 3.

    - Keep going until you hit 1.

In [42]:
# a function that adds previous numbers

def sum_prev(n):
    # base condition
    if n == 0:
        return n
    
    # problem to solve, and recursion
    return n + countdown(n-1)

sum_prev(10)

55

In [44]:
# a function that returns the factorial of a number

def factorialof(n):
    # case to help stop
    if n == 0:
        return 1 # if you keep n,then all numbers are multipled by 0 in end
    
    # return the problem to solve
    return n * factorialof(n-1)

factorialof(5)

120

6

Anything you can do with Recursion, you can do with loops too but:

- Some problems become much easier to think about with recursion, especially when:

    - a big problem can be cleanly broken into smaller same-shaped problems
(like trees, folders in folders, etc.)

<div style="background-color: lightblue; color: white; padding: 10px; text-align: center;">
    <h1>_________________________________END________________________________
</h1> </div>

<div class="alert alert-block alert-warning">
    <b><font size="5"> Live Exercise</font> </b>
</div>

Now it's your turn!
### Task: Create a Function to Calculate Factorial
Write a function named factorial that takes a positive integer n as input and returns its factorial. The factorial of a non-negative integer n is the product of all positive integers less than or equal to n. For example, the factorial of 5 (denoted as 5!) is 5 × 4 × 3 × 2 × 1 = 120.



### Mini Project – Calculator App

Build a console calculator that:

Asks the user for an operation: +, -, *, /

Asks for two numbers

Shows the result

Keeps running until user enters q to quit

<div style="background-color: #002147; color: #fff; padding: 30px; text-align: center;">
    <h1>THANK YOU!
</h1> </div>

<div style="background-color: lightgreen; color: black; padding: 30px;">
    <h4> Solution:
</h4> </div>

In [ ]:
# Simple Solution to exercise

def factorial(n):
    # This function calculates the factorial of a positive integer n
    result = 1  # Initialize the result variable to 1
    for i in range(1, n + 1):  # Loop from 1 to n (inclusive)
        result *= i  # Multiply result by the current number i
    return result  # Return the final factorial result

# Example usage
number = 5  # Set the number for which we want to calculate the factorial
print(f"The factorial of {number} is {factorial(number)}.")  # Output the result

In [ ]:
# Exact solution to exercise

def factorial(n):
    # This function calculates the factorial of a positive integer n
    if n < 0:
        return "Factorial is not defined for negative numbers"  # Check for negative input
    elif n == 0 or n == 1:
        return 1  # The factorial of 0 or 1 is 1
    else:
        result = 1  # Initialize a variable to hold the factorial result
        for i in range(2, n + 1):  # Loop from 2 to n (inclusive)
            result *= i  # Multiply result by the current number i
        return result  # Return the final factorial result

# Example usage
number = 5  # Set the number for which we want to calculate the factorial
factorial_result = factorial(number)  # Call the factorial function
print(f"The factorial of {number} is {factorial_result}.")  # Output the result


The factorial of 5 is 120.


- Mini project

In [ ]:
# hint implementation

def add(a, b):
    return a + b

def subtract(a, b):
    return a - b

def multiply(a, b):
    return a * b

def divide(a, b):
    if b == 0:
        print("Error: Cannot divide by zero!")
        return None
    return a / b

In [ ]:
# actual code

def calculator():
    print("Welcome to the Python Calculator 🧮")
    print("Operations: +  -  *  /")
    print("Type 'q' to quit.\n")

    while True:
        op = input("Choose operation (+, -, *, /) or q to quit: ").strip()

        if op == 'q':
            print("Goodbye! 👋")
            break

        if op not in ['+', '-', '*', '/']:
            print("Invalid operation. Try again.\n")
            continue

        # get numbers from user
        try:
            a = float(input("Enter first number: "))
            b = float(input("Enter second number: "))
        except ValueError:
            print("Please enter valid numbers!\n")
            continue

        # perform operation
        if op == '+':
            result = add(a, b)
        elif op == '-':
            result = subtract(a, b)
        elif op == '*':
            result = multiply(a, b)
        elif op == '/':
            result = divide(a, b)

        if result is not None:
            print("Result:", result, "\n")

# run calculator
calculator()


<div style="background-color: lightblue; color: white; padding: 10px; text-align: center;">
    <h1>_________________________________END________________________________
</h1> </div>

-----------------------

-----------------------

-----------------------

-----------------------

<div style="background-color: lightgreen; color: black; padding: 4px;">
    <h4>1. Imperative Programming
</h4> </div>

**Definition**: This paradigm focuses on commands that change a program’s state. It involves writing sequences of statements that lead to a desired outcome.

**Characteristics**: 
  - Uses statements to change program state.
  - Follows a step-by-step approach.

In [11]:
# Example
total = 0
for i in range(1, 6):
    total += i  # Add i to total
print(total)  # Output: 15


15


<div style="background-color: lightgreen; color: black; padding: 4px;">
    <h4>2. Procedural Programming
</h4> </div>

**Definition**: A subtype of imperative programming that structures code into procedures (or functions). It promotes code reuse and organization.

**Characteristics**: 
  - Uses functions to encapsulate behavior.
  - Supports modularity.

In [17]:
#example 

def calculate_sum(n):
    total = 0
    for i in range(1, n + 1):
        total += i
    return total

print(calculate_sum(5))  # Output: 15


15


<div style="background-color: lightgreen; color: black; padding: 4px;">
    <h4>3. Object-Oriented Programming (OOP)
</h4> </div>

**Definition**: This paradigm is based on the concept of "objects," which are instances of classes. OOP promotes encapsulation, inheritance, and polymorphism.

**Characteristics**: 
  - Encapsulation: Bundling data and methods that operate on the data within one unit.
  - Inheritance: Mechanism to create a new class using the properties of an existing class.
  - Polymorphism: Ability to use a single interface for different data types.

In [21]:
# Example
class Dog:
    def __init__(self, name):
        self.name = name
    
    def bark(self):
        return f"{self.name} says woof!"

my_dog = Dog("Buddy")
print(my_dog.bark())  # Output: Buddy says woof!


Buddy says woof!


<div style="background-color: lightgreen; color: black; padding: 4px;">
    <h4>4. Functional Programming
</h4> </div>

**Definition**: This paradigm treats computation as the evaluation of mathematical functions and avoids changing state or mutable data.

**Characteristics**: 
  - First-class functions: Functions can be passed as arguments, returned from other functions, and assigned to variables.
  - Higher-order functions: Functions that operate on other functions. 
  - Immutability: Data is immutable.

In [31]:
# Example
def square(x):
    return x * x

numbers = [1, 2, 3, 4]
squares = list(map(square, numbers))  # Apply the square function to each element
print(squares)  # Output: [1, 4, 9, 16]


      Name  Score
2  Charlie     95


<div style="background-color: lightgreen; color: black; padding: 4px;">
    <h4>5. Declarative Programming
</h4> </div>

**Definition**: This paradigm expresses the logic of a computation without describing its control flow. It focuses on what the program should accomplish rather than how to achieve it.

**Characteristics**: 
  - Focuses on the outcome rather than the procedure.
  - Commonly used in database querying and configuration management.

**Example:**
  - SQL for querying databases is a common example, while in Python, you can use libraries like Pandas to filter data in a declarative manner.


In [ ]:
#Example
import pandas as pd

data = {'Name': ['Alice', 'Bob', 'Charlie'],
        'Score': [85, 90, 95]}
df = pd.DataFrame(data)

# Filter where Score is greater than 90
high_scores = df[df['Score'] > 90]
print(high_scores)
